### ISSUE TITLE
Re-write Authentication

### ISSUE THREAD

**Issue by Lange** [nodecg/nodecg#139]
NodeCG's current authentication is terrible. It's hardcoded to only support Twitch and Steam as auth providers and has many severe bugs such as the infamous "poison cookie" bug, which prevents user login until they clear their cookies.

The entire auth system needs to be thrown out and redesigned. Out-of-the-box it should support local auth. That is to say, normal username/password auth directly into the NodeCG instance.

I'm not sure how to best support other auth providers such as Steam and Twitch. Ideally this auth system would be extensible, so bundle authors could provide new auth mechanisms without making a PR to NodeCG. I've no idea how to accomplish this in a clean way, nor if it's actually even the best route to take.

**Comment by ProbablePrime** [comment]
[NodeBB](https://nodebb.org/) appears to allow other developers to write auth plugins that hook into nodebb somehow.

I haven't done any indepth research but mentioning it as something that might be worth a bit of reading to help get the ball rolling.

[NodeBB Plugin Information](https://docs.nodebb.org/en/latest/plugins/create.html?highlight=plugins)

Most of them appear to use passport within the plugin which nodecg is already using.

You can find a list of their [auth plugins on npm](https://www.npmjs.com/search?q=nodebb-plugin-sso-)

**Comment by MattMcNam** [comment]
I've been wanting to try something similar to Atom's plugins, and NodeBB's too by the looks of it, in a bundle, but it would fit in here to.

Export functions in a module, map those to hooks in package.json

**Comment by Lange** [comment]
@MattMcNam Would that look something like this?

```js
module.exports = {
    // An array allows multiple authProviders to be defined.
    authProviders: [{
        name: 'google',
        strategySetup: function (passport, config) {
            var GoogleStrategy = require('passport-google-strategy');
            passport.use(new GoogleStrategy({
                foo: 'foo',
                bar: 'bar'
            }), function (profile, done) {
                profile.allowed = (config.login.google.allowedEmails.indexOf(profile.email) > -1);

                if (profile.allowed) {
                    log.info('Granting %s access', profile.email, profile.displayName);
                } else {
                    log.info('Denying %s access', profile.email, profile.displayName);
                }

                return done(null, profile);
            });
        }
    }]

    // Other plugin hooks can go here?
};
```

Where would this plugin file go? Should plugins be a separate concept from extensions? Probably not, no reason to have two similar and confusing concepts. How can the extension format be changed to make these types of plugins possible?

**Comment by Lange** [comment]
While I do want to undergo a full-scale re-write of Auth, I think there's room for some incremental improvements that don't require breaking changes. I'm going to explore the possibilities of improving the current system in `0.7.x`.

**Comment by Lange** [comment]
I just merged #148 which I believe fixes the existing bugs in the login system. It's no longer a flaming pile of trash now.

I'm still leaning towards more incremental improvements before a full pluggable re-write. I think there's a lot more that the current system can do, and anything I can do to avoid having to spend days designing an entirely new system is good.

**Comment by Lange** [comment]
This issue is pretty stale, and has nothing actionable. Closing.
